edge-hub/watchdog/src/main.rs

```diff
@@ -0,0 +1,82 @@
+mod child;
+
+use std::{
+    io::Error,
+    sync::atomic::{AtomicBool, Ordering},
+    sync::Arc,
+    thread,
+};
+
+use anyhow::{Context, Result};
+use child::run;
+use signal_hook::{iterator::Signals, SIGINT, SIGTERM};
+use tracing::{error, info, subscriber, Level};
+use tracing_subscriber::fmt::Subscriber;
+
+fn main() -> Result<()> {
+    init_logging();
+    info!("Starting Watchdog");
+
+    let should_shutdown = register_shutdown_listener()
+        .context("Failed to register sigterm listener. Shutting down.")?;
+
+    let edgehub_handle = run(
+        "Edge Hub".to_string(),
+        "dotnet".to_string(),
+        "/app/Microsoft.Azure.Devices.Edge.Hub.Service.dll".to_string(),
+        Arc::clone(&should_shutdown),
+    )?;
+
+    let broker_handle = match run(
+        "MQTT Broker".to_string(),
+        "/usr/local/bin/mqttd".to_string(),
+        "-c /tmp/mqtt/config/production.json".to_string(),
+        Arc::clone(&should_shutdown),
+    ) {
+        Ok(handle) => Some(handle),
+        Err(e) => {
+            should_shutdown.store(true, Ordering::Relaxed);
+            error!("Could not start MQTT Broker process. {}", e);
+            None
+        }
+    };
+
+    if let Err(e) = edgehub_handle.join() {
+        should_shutdown.store(true, Ordering::Relaxed);
+        error!("Failure while running MQTT Broker process. {:?}", e)
+    }
+    info!("Successfully stopped Edge Hub process");
+
+    if let Some(Err(e)) = broker_handle.map(|handle| handle.join()) {
+        should_shutdown.store(true, Ordering::Relaxed);
+        error!("Failure while running MQTT Broker process. {:?}", e);
```

Comment:
I think you need to stop here. otherwise, log messages will confuse.
```suggestion
        bail!("Failure while running MQTT Broker process. {:?}", e);
```

Reply:
I think I would rather fix this with else

Reply:
The case on line 44 is easier to handle. For this block I thought I could add an ```else if let Some(_) = broker_handle``` but I don't think this is true because ```broker_handle``` is moved in the initial map() call

Reply:
I used a pattern shown here:
https://doc.rust-lang.org/stable/rust-by-example/flow_control/if_let.html

See ```// Provide an altered failing condition```

Reply:
yeah, the second block is a bit harder to read though?
let's go back with 
```rust
if let Some(handle) = broker_handle {
  if let Err(e) = handle.join() {
    ...
  }
} else {
  ...
}
```
another fancy way to solve nesting problem is `transpose` operator
```rust
if let Some(_) = broker_handle
  .map(|handle|handle.join().map_err(|e| anyhow!("Failure while running MQTT Broker process. {:?}", e)) // makes Option<Result<()>>
  .transpose()? { // makes Result<Option<()>>
  info!("Successfully stopped MQTT Broker process");
}  
```

Reply:
I was playing around with the transpose, and it was autoformatting to more nested levels than your first block suggestion above :).

I'm going to go with the second suggestion, but I think we want the ending else statement to be inside the top-level if block.